crates/iceberg/src/writer/file_writer/location_generator.rs

```diff
@@ -0,0 +1,108 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains the location generator and file name generator for generating path of data file.
+
+use std::sync::{atomic::AtomicU64, Arc};
+
+/// `LocationGenerator` used to generate the location of data file.
+pub trait LocationGenerator: Clone + Send + 'static {
+    /// Generate a absolute path for the given file name.
+    /// e.g
+    /// For file name "part-00000.parquet", the generated location maybe "/table/data/part-00000.parquet"
+    fn generate_location(&self, file_name: &str) -> String;
+}
+
+/// `FileNameGeneratorTrait` used to generate file name for data file. The file name can be passed to `LocationGenerator` to generate the location of the file.
+pub trait FileNameGenerator: Clone + Send + 'static {
+    /// Generate a file name.
+    fn generate_file_name(&self) -> String;
+}
+
+/// `DefaultFileNameGenerator` used to generate file name for data file. The file name can be
+/// passed to `LocationGenerator` to generate the location of the file.
+/// The rule of file name is aligned with the OutputFileFactory in iceberg-java
+#[derive(Clone)]
+pub struct DefaultFileNameGenerator {
+    partition_id: u64,
```

Comment:
I'm hesitating to add these concepts `partition_id`, `task_id`, `operator_id` in the core crate, since these are all computation related concepts. As default, I this a random uuid would be enough. Compute engines can always override this trait.

Reply:
Sounds reasonable for me.